### PR TITLE
Feature/merge sage maker outputs

### PR DESCRIPTION
**Pull request by djkapner**
### Overview
Adds utility functions to merge multiple SagemakerGroundTruth output manifests into 1 manifest. This does not have a CLI, as it is expected to be used during data-science mode to prepare some classifier inputs. An example of doing that is provided in this description. This is the machinery for #25, and this PR demonstrates that it is ready as soon as the annotators are done.

### Validation
The following call retrieves 2 output manifests from S3 and merges them (the code in this PR) and then talks to on-prem databases to assemble the training data (not part of this PR, uses DB connections from this repo). The code for this file is included here, collapsed, as a starting point example for the next dev starting to cook up a jupyter notebook or something.
<details>
  <summary>example_merge_and_prepare.py</summary>

```python
import logging
import json
import numpy as np

import slapp.utils.merge_utils as mu
from slapp.utils import query_utils
from slapp.rois import ROI

logging.getLogger().setLevel(logging.INFO)


def meta_from_experiment_id(experiment_id, db_connection):
    meta_data = db_connection.query(
        "SELECT e.name AS rig, i.depth, st.acronym AS targeted_structure, "
        "d.full_genotype FROM ophys_experiments oe "
        "JOIN ophys_sessions os ON oe.ophys_session_id=os.id "
        "JOIN equipment e ON e.id=os.equipment_id "
        "JOIN specimens sp ON sp.id=os.specimen_id "
        "JOIN imaging_depths i ON i.id=oe.imaging_depth_id "
        "JOIN structures st ON st.id=oe.targeted_structure_id "
        "JOIN donors d ON d.id=sp.donor_id "
        f"WHERE oe.id = {experiment_id}")

    if len(meta_data) != 1:
        raise ValueError(f"expected 1 record for {experiment_id}, "
                         f"found {len(meta_data)}")

    return meta_data[0]


# merge some output manifests together
src_uris = [
        "s3://prod.slapp.alleninstitute.org/2-line 2-project/20200520051838/2-line-2-project-prod-v2-output/2-line-2-project-prod-v2/manifests/output/output.manifest",
        "s3://prod.slapp.alleninstitute.org/2-line 2-project/20200520051838/2-line-2-project-output/2-line-2-project-prod/manifests/output/output.manifest"]

dst_uri = "/home/danielk/sgt_merged_manifest.jsonl"

_ = mu.merge_outputs(src_uris, dst_uri)

# select only the valid ones
x = list(mu.read_jsonlines(dst_uri))
valid = [i for i in x if i['merged-project']['majorityLabel'] is not None]

# retrieve the data necessary for input to
label_credentials = query_utils.get_db_credentials(
        env_prefix="LABELING_",
        **query_utils.label_defaults)
label_connection = query_utils.DbConnection(**label_credentials)
lims_credentials = query_utils.get_db_credentials(
        env_prefix="LIMS_",
        **query_utils.lims_defaults)
lims_connection = query_utils.DbConnection(**lims_credentials)

rois = []
data = []
translator = {'cell': 1, 'not cell': 0}
for record in valid:
    roi = ROI.roi_from_query(record['roi-id'], label_connection)
    meta = meta_from_experiment_id(record['experiment-id'], lims_connection)
    entry = {
            "roi_id": roi.roi_id,
            "coo_cols": roi._sparse_coo.col.tolist(),
            "coo_rows": roi._sparse_coo.row.tolist(),
            "coo_data": roi._sparse_coo.data.tolist(),
            "image_shape": roi.image_shape,
            "trace": roi.trace
            }
    entry.update(meta)
    entry.update(
            {"label": translator[record['merged-project']['majorityLabel']]})
    data.append(entry)


# train test split (scikit-learn not a dependency of slapp)
def train_test_split(data, test_size, random_state=0):
    ntest = int(test_size * len(data))
    rng = np.random.default_rng(random_state)
    rng.shuffle(data)
    return data[ntest:], data[:ntest]


train, test = train_test_split(data, test_size=0.3, random_state=0)

training_path = "/home/danielk/mlflow_example/training_data.json"
with open(training_path, "w") as f:
    json.dump(train, f)
testing_path = "/home/danielk/mlflow_example/testing_data.json"
with open(testing_path, "w") as f:
    json.dump(test, f)
print(f"wrote {training_path}")
print(f"wrote {testing_path}")

```

</details>

```
$ python example_merge_and_prepare.py
WARNING:root:skipped some records {
  "s3://prod.slapp.alleninstitute.org/2-line 2-project/20200520051838/2-line-2-project-prod-v2-output/2-line-2-project-prod-v2/manifests/output/output.manifest": 392,
  "s3://prod.slapp.alleninstitute.org/2-line 2-project/20200520051838/2-line-2-project-output/2-line-2-project-prod/manifests/output/output.manifest": 765
}
INFO:root:n records used: {
  "s3://prod.slapp.alleninstitute.org/2-line 2-project/20200520051838/2-line-2-project-prod-v2-output/2-line-2-project-prod-v2/manifests/output/output.manifest": 608,
  "s3://prod.slapp.alleninstitute.org/2-line 2-project/20200520051838/2-line-2-project-output/2-line-2-project-prod/manifests/output/output.manifest": 235
}
INFO:root:wrote /home/danielk/sgt_merged_manifest.jsonl with 694 records and 198 valid majorities.
wrote /home/danielk/mlflow_example/training_data.json
wrote /home/danielk/mlflow_example/testing_data.json
```
The inputs for training have now been created. 
Then, following the [automated mlflow CLI test](https://github.com/AllenInstitute/croissant/blob/01f580bfb80a411f47bf285ff29b0620548b2dcb/.circleci/config.yml#L57-L71) as an example.
```
export MLFLOW_TRACKING_URI=/home/danielk/mlflow_example/tracking
export ARTIFACT_URI=/home/danielk/mlflow_example/artifacts
export MYEXPERIMENT=my_experiment
mlflow experiments create \
              --experiment-name ${MYEXPERIMENT} \
              --artifact-location ${ARTIFACT_URI}
```
The mlflow run command is being run from the `croissant` repo root dir to pass in the `MLProject` file.
```
~/croissant$ mlflow run . --backend local --experiment-name ${MYEXPERIMENT} -P training_data=/home/danielk/mlflow_example/training_data.json -P test_data=/home/danielk/mlflow_example/testing_data.json -P scoring="['roc_auc']" -P refit=roc_auc -P log_level=INFO

2020/07/09 15:57:59 INFO mlflow.projects: === Created directory /tmp/tmps19ei_lc for downloading remote URIs passed to arguments of type 'path' ===
2020/07/09 15:57:59 INFO mlflow.projects: === Running command 'source /home/danielk/miniconda3/bin/../etc/profile.d/conda.sh && conda activate mlflow-ead6525c1aa4af043de9afdf421d4abe083b57fe 1>&2 && python -m croissant.train --training_data /home/danielk/mlflow_example/training_data.json --test_data /home/danielk/mlflow_example/testing_data.json --scoring '['"'"'roc_auc'"'"']' --refit roc_auc --log_level INFO' in run with ID '5f5b74552cea4e08b626012146bb1ad3' === 
INFO:TrainClassifier:Reading training data and extracting features.
INFO:TrainClassifier:Fitting model to data!
INFO:TrainClassifier:fitting model with **<snipped>**
**<snipped convergence warnings>**
INFO:ClassifierTrainer:Model fit, with best score 0.6386397707231041 and best parameters {'model__l1_ratio': 0.25}.
/home/danielk/miniconda3/envs/mlflow-ead6525c1aa4af043de9afdf421d4abe083b57fe/lib/python3.8/site-packages/joblib/numpy_pickle.py:103: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  pickler.file_handle.write(chunk.tostring('C'))
INFO:ClassifierTrainer:logged training to mlflow run 5f5b74552cea4e08b626012146bb1ad3
2020/07/09 15:58:05 INFO mlflow.projects: === Run (ID '5f5b74552cea4e08b626012146bb1ad3') succeeded ===

```
which has written a model:
```
$ ls /home/danielk/mlflow_example/artifacts/5f5b74552cea4e08b626012146bb1ad3/artifacts/
trained_model.joblib
```
and we can also confirm this through the mlflow ui:
```
$ mlflow ui --backend-store-uri /home/danielk/mlflow_example/tracking
```
image

### CODE REVIEW

File: slapp/transfers/utils.py [hunk 1]
```diff
@@ -2,10 +2,50 @@
...
+
+
+def read_jsonlines(
+        uri: Union[str, pathlib.Path]) -> Generator[dict, None, None]:
+    """
+    Generator to load jsonlines file from either s3 or a local
+    file, given a uri (s3 uri or local filepath).
+    """
+    if str(uri).startswith("s3://"):
+        data = s3_get_object(uri)["Body"].iter_lines(chunk_size=8192)    # The lines can be big        # noqa
```
Comment (njmei) [inline slapp/transfers/utils.py:41]: why `# noqa`? This could mask issues if someone changes this line in the future. If it's due to line length just put the comment one line above?

File: slapp/utils/merge_utils.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+class MergingException(Exception):
+    pass
+
+
+class WorkerAnnotation(TypedDict):
+    workerId: str
+    roiLabel: str
+
+
+class RecordProject(TypedDict):
```
Comment (njmei) [inline slapp/utils/merge_utils.py:24]: Awesome idea to start defining `TypedDict`s!

File: slapp/utils/merge_utils.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    sourceData: str
+    majorityLabel: str
+    workerAnnotations: List[WorkerAnnotation]
+
+
+def compute_majority(labels):
+    """return majority labels or None if tied
```
Comment (njmei) [inline slapp/utils/merge_utils.py:31]: Could you add more information (and typing) about how `labels` is organized?

File: slapp/utils/merge_utils.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    ulabels = set(labels)
+    if len(ulabels) not in [1, 2]:
+        raise ValueError("`compute_majority()` expects binary classification "
+                         f"but received {len(ulabels)} labels: {ulabels}")
+    counts = {ulabel: labels.count(ulabel) for ulabel in ulabels}
```
Comment (njmei) [inline slapp/utils/merge_utils.py:37]: This is fine, you could replace it with a [`collections.Counter`](https://docs.python.org/3.8/library/collections.html#collections.Counter) I think and not even need to create `ulabels` I think?

File: slapp/utils/merge_utils.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    # require 3 and only 3 annotations
+    if sum(counts.values()) == 3:
+        mcount = max(counts.values())
+        r_counts = {v: k for k, v in counts.items()}
+        majority = r_counts[mcount]
+
+    return majority
+
+
+def get_record_project_key(record: dict):
```
Comment (njmei) [inline slapp/utils/merge_utils.py:50]: Perhaps `record` should also have a `TypedDict` entry?

Comment (djkapner) [reply slapp/utils/merge_utils.py:50]: I wanted to do that, but here's the catch. The record keys `myproject` and `myproject-metadata` are different for every launched job. So, we don't know a-priori their names. Can you think of a solution?

File: slapp/transfers/utils.py (hunk 1 above)
Comment (djkapner) [reply slapp/transfers/utils.py]: done

File: slapp/utils/merge_utils.py (hunk 5 above)
Comment (njmei) [reply slapp/utils/merge_utils.py:50]: Maybe setting `total=False` and making a note about the undocumented `myproject` and `myproject-metadata` keys?

https://mypy.readthedocs.io/en/latest/more_types.html#totality

File: slapp/utils/merge_utils.py (hunk 3 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py:31]: I'll add docstrings... sorry about that

File: slapp/utils/merge_utils.py (hunk 5 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py:50]: right, but the hierarchy of this is:
```
Record - |
               | - RecordProject - |
                                             | - WorkerAnnotations
```
That's what would be useful, but for this:
```
wa = myrecord['myproject']['WorkerAnnotations']
```
`myproject` is dynamic, so I can't specify a TypedDict like:
```
class Record(TypedDict):
    myproject: RecordProject
```
because I don't know the name <myproject>

File: slapp/utils/merge_utils.py (hunk 4 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py:50]: I did that at first, and I didn't find it much better. I could have gone either way

File: slapp/utils/merge_utils.py (hunk 3 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py]: done

File: slapp/utils/merge_utils.py (hunk 5 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py]: I think a way to do this is :
```
class ParsedRecord(dict):
    @classmethod
    def from_record(cls, dict_in: dict):
        d = {}
        pkey = get_project_key(dict_in)
        if pkey is None:
            return None
        d['project-metadata'] = dict_in.pop(pkey + '-metadata')
        d['project'] = Project(dict_in.pop(pkey))
        d['project_key'] = pkey
        for k, v in dict_in.items():
            d[k] = v
        print(json.dumps(d, indent=2))
        return ParsedRecord(**d)

    def to_record(self):
        pkey = self.pop('project_key')
        self[pkey] = self.pop('project')
        self[pkey + '-metadata'] = self.pop('project-metadata')
```
but as we discussed offline, this is more work than we need right now. Maybe we can come back to this later if we want.

File: slapp/utils/merge_utils.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    ----------
+    labels: list
+        items could be for example "cell"/"not cell" or 1/0
+        our SagemakerGroundTruth jobs are using the strings
+
+    Returns
+    -------
+    majority:
+       element of labels that is most prevalent or None if
+       lenght of labels is not 3
```
Comment (kschelonka) [inline slapp/utils/merge_utils.py:43]: 👍 

Typo but nbd

File: slapp/utils/merge_utils.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    # require 3 and only 3 annotations
+    if sum(counts.values()) == 3:
+        mcount = max(counts.values())
+        r_counts = {v: k for k, v in counts.items()}
+        majority = r_counts[mcount]
+
+    return majority
+
+
+def get_project_key(record: dict) -> Union[str, None]:
```
Comment (kschelonka) [inline slapp/utils/merge_utils.py:63]: I'd make a note in the docstring that this looks for an additional key in the data that ends with "-metadata", then strips the "-metadata" to get the project_key. If SageMaker changes the way they format the output then this function will break, so it will be easier to see if we record it in the docstring.

File: slapp/utils/merge_utils.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    Returns
+    -------
+    new_project: Project
+        new with workerAnnotations concatenated and sourceData appended
+        if different. Any other entities inherited from project1
+
+    """
+    new_project = Project(copy.deepcopy(project1))
+    new_project['workerAnnotations'].extend(project2['workerAnnotations'])
+    if new_project['sourceData'] != project2['sourceData']:
```
Comment (kschelonka) [inline slapp/utils/merge_utils.py:119]: Might be clearer if this is comma-delimited rather than space, to concatenate these URIs. Did you find this case in the data?

File: tests/utils/test_merge_utils.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+from moto import mock_s3
+
+import slapp.utils.merge_utils as mu
+
+
+@pytest.mark.parametrize(
+        "labels, expected",
+        [
+            # not 3 annotations
+            ([0, 0, 0, 1, 1], None),
```
Comment (kschelonka) [inline tests/utils/test_merge_utils.py:13]: Less than 3?

Comment (djkapner) [reply tests/utils/test_merge_utils.py]: outdated, using the post-annotation compute_majority now

File: slapp/utils/merge_utils.py (hunk 8 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py:87]: done.
No, but I could imagine a scenario where it would happen. Hopefully not.

File: slapp/utils/merge_utils.py (hunk 6 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py]: outdated, using different function

File: slapp/utils/merge_utils.py (hunk 7 above)
Comment (djkapner) [reply slapp/utils/merge_utils.py:31]: done

**Comment by djkapner** [comment]
> LGTM overall. Couple minor notes mostly for clarification.

@kschelonka updated with requested changes and using post-lambda compute_majority. Reran the example in the description and confirmed it works.
